Ports aiming component from NSV13

benbot16 opened the pull request:
<!-- Write **BELOW** The Headers and **ABOVE** The comments else it may not be viewable. -->
<!-- You can view Contributing.MD for a detailed description of the pull request process. -->

## About The Pull Request

<!-- Describe The Pull Request. Please be sure every change is documented or this can delay review and even discourage maintainers from merging your PR! -->
This PR takes the aiming component added to NSV13 in BeeStation/NSV13#607 and ports it. 

In a nutshell, this adds the ability to aim weapons at players by pointing at them. The aimer and target then both get popup menus. This menu allows the aimer to shoot the target or shout a command at them, and gives the target an option to surrender. The aimer will also get notifications if the target tries to equip/drop an object.

Compared to NSV's version, I removed the stamina damage buffs to aimed bullets and removed the aiming ability button. The first was done because it's a bit too OP, and the second was done since it clutters up the ability bar too much. I also added a check to prevent people from aiming/firing while not in view of the target.

I also added checks to autofire to prevent it from firing when trying to aim at/examine someone, so that holding up people with an automatic rifle won't immediately shoot them.

## Why It's Good For The Game

<!-- Please add a short description of why you think these changes would benefit the game. If you can't justify it in words, it might not be worth adding. -->
This makes holding someone at gunpoint more viable than it would be otherwise, especially when using lethal weapons.

## Changelog
:cl:
add: You can now hold people at gunpoint by pointing at them while holding a gun.
fix: Fixes automatic weapons shooting when examining, pointing, or clicking certain UI items.
/:cl:

<!-- Both :cl:'s are required for the changelog to work! You can put your name to the right of the first :cl: if you want to overwrite your GitHub username as author ingame. -->
<!-- You can use multiple of the same prefix (they're only used for the icon ingame) and delete the unneeded ones. Despite some of the tags, changelogs should generally represent how a player might be affected by the changes rather than a summary of the PR's contents. -->

benbot16 committed:
Auto stash before merge of "master" and "BeeStation/master"

benbot16 committed:
pain and suffering

benbot16 committed:
Adds aiming component

benbot16 committed:
View check for shooting at people you're aiming at

benbot16 committed:
change to waving and fix for autofire

benbot16 committed:
Better way of fixing autofire being fucky

benbot16 committed:
Removes something unused

benbot16 committed:
Some changes to displayed text

benbot16 committed:
Fixing autofire examining while I'm here

benbot16 committed:
Adds a couple actions to some of the aiming stages

Kapu1178 commented:
I was going to port this from bay but this seems like an infinitely better port. NICE

Kapu1178 commented on code/datums/components/aiming.dm:
```diff
@@ -0,0 +1,168 @@
+	user.visible_message("<span class='warning'>[user] points [parent] at [target]!</span>")
+	to_chat(target, "<span class='userdanger'>[user] is pointing [parent] at you! If you equip or drop anything they will be notified! \n <b>You can use *surrender to give yourself up</b>.</span>")
+	to_chat(user, "<span class='notice'>You're now aiming at [target]. If they attempt to equip anything you'll be notified by a loud sound.</span>")
+	user.say("FREEZE!!")
```
```suggestion
```

Players can say this themselves. Forced speech bad.

Commit:
Update code/datums/components/aiming.dm

Co-authored-by: Kapu1178 <75460809+Kapu1178@users.noreply.github.com>

Archanial commented on code/datums/components/aiming.dm:
```diff
@@ -0,0 +1,167 @@
+	var/mob/living/user = null
+	var/mob/living/target = null
+	var/aiming_cooldown = FALSE
+	var/cooldown_time = 5 SECONDS // So you can't spam aiming for faster bullets/spamming lines
```
Use our cooldown macros instead

benbot16 replied:
Changed it to use the macros, let me know if I forgot anything.

benbot16 committed:
Changes aiming cooldown to use the proper macros

Kapu1178 commented on code/datums/components/aiming.dm:
```diff
@@ -0,0 +1,165 @@
+		if("face_wall")
+			user.say("TURN AROUND AND FACE THE WALL. SLOWLY.")
+		if("drop_to_floor")
+			user.say("ON THE FLOOR, NOW!")
```
forced speech bad.
```suggestion
```

benbot16 replied:
This isn't something you actively have to use though, it's more so people can give quick commands if they wouldn't otherwise have time to type something out.

Kapu1178 replied:
Eeeeeeehhhh fine

ta3370 replied:
Does this have a Cooldown? And is it mime-proof if they get the "fingergun"? There was an issue on NSV a while back where people spammed the text buttons.

Other than that, glad to see this coming to Bee, it opens up much more opportunity to have an interim where somebody knows you're aiming a weapon at them. Many options for player interaction opening up.

benbot16 replied:
Didn't have either of those, but I added them. Dunno about the cooldown length, but I can change that later if it's too much/little.

TzuInferno replied:
Honestly this always felt kind of weird to me where commands were added or randomly said? Why not just make it "click yell" "click what you want to yell"?

benbot16 replied:
that's how it works, there's a button that corresponds to each command. When you click it, you say that command.

TzuInferno replied:
Oh, I'm dumb then!

Archanial replied:
Instead of force say make use of balloon alerts and display something like "[x] commands you to [y]".
Don't forget to display balloon alert for viewers with appropriate message

Archanial commented on code/datums/components/aiming.dm:
```diff
@@ -0,0 +1,165 @@
+	user.visible_message("<span class='warning'>[user] points [parent] at [target]!</span>")
+	to_chat(target, "<span class='userdanger'>[user] is pointing [parent] at you! If you equip or drop anything they will be notified! \n <b>You can use *surrender to give yourself up</b>.</span>")
+	to_chat(user, "<span class='notice'>You're now aiming at [target]. If they attempt to equip anything you'll be notified by a loud sound.</span>")
+	playsound(target, 'sound/weapons/autoguninsert.ogg', 100, TRUE)
```
This could use some balloon message to target

benbot16 replied:
I put one on the target, although it feels a bit redundant since the target already gets a radial popup and crosshairs over them

Archanial commented on code/datums/components/aiming.dm:
```diff
@@ -0,0 +1,165 @@
+*/
+
+/datum/component/aiming/proc/on_drop()
+	to_chat(user, "<span class='nicegreen'>[target] has dropped something.</span>")
```
Missing signal handlers

benbot16 replied:
Added them

benbot16 committed:
Adds some signal handlers and a balloon_alert

Raven-Industries commented:
testmerge when?

benbot16 committed:
Adds a cooldown for voice lines

Also adds special handling for mimes, and forced tags on the speech (because I forgot to add them earlier)

heepox commented:
Finally some good fucking PR

Time to hold people up with fingerguns.

TzuInferno commented:
testmerge when

ta3370 commented:
Will this be getting Testmerged?

ike709 commented:
You all have one (1) pass to annoy one (1) maintainer that isn't me (since I am physically unable to testmerge for a while) on discord.

Archanial commented on code/datums/components/aiming.dm:
```diff
@@ -0,0 +1,181 @@
+	COOLDOWN_DECLARE(voiceline_cooldown)
+
+/datum/component/aiming/Initialize(source)
+	. = ..()
```
parent proc is empty

Archanial commented on code/datums/components/aiming.dm:
```diff
@@ -0,0 +1,181 @@
+	user.visible_message("<span class='warning'>[user] points [parent] at [target]!</span>")
+	to_chat(target, "<span class='userdanger'>[user] is pointing [parent] at you! If you equip or drop anything they will be notified! \n <b>You can use *surrender to give yourself up</b>.</span>")
+	to_chat(user, "<span class='notice'>You're now aiming at [target]. If they attempt to equip anything you'll be notified by a loud sound.</span>")
+	target.balloon_alert(target, "[user] is pointing [parent] at you!")
```
You should use balloon alert to viewers

Archanial commented on code/datums/components/aiming.dm:
```diff
@@ -0,0 +1,181 @@
+*/
+
+/datum/component/aiming/proc/on_drop()
+	SIGNAL_HANDLER
```
Generally I feel like there's a lot of places where you could add balloon alerts

Archanial commented on code/modules/projectiles/gun.dm:
```diff
@@ -635,8 +635,10 @@
-	return
+/obj/item/gun/proc/before_firing(atom/target, mob/user, aimed)
+	if(aimed)
+		if(chambered?.BB && !istype(src, /obj/item/gun/ballistic/automatic/toy))
```
Instead of snowflake check for toys just override proc for them

benbot16 committed:
Yet more balloon alerts plus reworking menus

- Changes most says to use balloon alerts
- Reworks the radial menus to auto-close when aiming stops
- Aiming now stops if the user and target move away or lose sight of each other
- adds a few comments because comments are nice
- Removes some snowflake code on guns when aiming

Archanial commented on code/datums/components/aiming.dm:
```diff
@@ -45,17 +51,27 @@ Methods to alert the aimer about events, usually to signify that they're complyi
 	SEND_SOUND(user, 'sound/machines/chime.ogg')
 
+// Cancels aiming if we can't see the target
+/datum/component/aiming/proc/on_move()
```
Signal handler

benbot16 committed:
Signal handler

benbot16 committed:
Parentheses

Archanial commented on code/modules/projectiles/gun.dm:
```diff
@@ -635,8 +635,10 @@
+/obj/item/gun/proc/before_firing(atom/target, mob/user, aimed)
+	if(aimed)
+		if(chambered?.BB)
+			chambered.BB.jitter = initial(chambered.BB.speed) *= 0.5 // Faster bullets to account for the fact you've given the target a big warning they're about to be shot
```
```suggestion
	if(aimed && chambered?.BB)
		chambered.BB.jitter = initial(chambered.BB.speed) *= 0.5 // Faster bullets to account for the fact you've given the target a big warning they're about to be shot
```

Archanial commented:
https://cdn.discordapp.com/attachments/565330972707651584/947512255442997308/unknown.png
https://media.discordapp.net/attachments/565330972707651584/947512364125814824/unknown.png
https://media.discordapp.net/attachments/565330972707651584/947513160699617320/unknown.png
https://cdn.discordapp.com/attachments/565330972707651584/947508864918622270/unknown.png

Guns don't shoot after being reloaded, they act as if they're empty

Commit:
Update code/modules/projectiles/gun.dm

Co-authored-by: Archanial <37456678+Archanial@users.noreply.github.com>

benbot16 committed:
A couple code changes

Moves components around a little bit and tweaks a couple of the user/target checks.

benbot16 committed:
pain

benbot16 committed:
Fixes autofire runtimes involving turfs

This should also fix the issue where you can't use the tiles next to you to fire when autofiring.

benbot16 committed:
Changes aiming speed buff a bit and adds a damage modifier

benbot16 committed:
Switches aim handling from using pointed() to using a point signal

This should fix the GetComponent() runtimes, but I haven't been able to replicate the SendSignal() runtimes, so I don't know if this'll do anything for those.

benbot16 commented:
This *should* fix the runtime in pointing, but I still haven't been able to find what was causing SendSignal to runtime.

benbot16 committed:
Active hand check for point signal handler

benbot16 committed:
Adds some defines to replace strings and a couple user/target checks

benbot16 committed:
Fixes radials not cleaning up and causing runtimes during aiming

This *might* break something if there's a radial somewhere that needs to exist without a menu, but I don't think any of those exist.

benbot16 committed:
Makes equip/drop notifications actually work

So when I originally made the port, these were actually broken. Apparently I never noticed that until now while I was testing. Anyway, it works now. The signals might need a look over, though.

Archanial commented:
![obraz](https://user-images.githubusercontent.com/37456678/163679691-1a8cfa3a-c365-478d-b845-7541abd96d71.png)
![obraz](https://user-images.githubusercontent.com/37456678/163679730-509d1615-3828-4e13-a306-97e8b0532055.png)

benbot16 committed:
Fixes a null invoke with persistent radials

There's a potential for radials to get here without a callback, so this just adds a check to make sure there's actually a callback.

PowerfulBacon commented on code/_onclick/hud/radial.dm:
```diff
@@ -18,7 +18,7 @@ GLOBAL_LIST_EMPTY(radial_menus)
 /atom/movable/screen/radial/proc/handle_parent_del()
 	SIGNAL_HANDLER
-	set_parent(null)
+	qdel(src) // No reason for us to exist if our parent menu's gone and we don't have a new one
```
Results in extremely intense hard dels

benbot16 replied:
Ah, didn't see this happening when I tested it. I'll try something else to fix the issue then.
